### PR TITLE
Extension system

### PR DESCRIPTION
**Pull request by forgee**
Still much left to be done but base functionality is there for pages/widgets.

### CODE REVIEW

**Comment by Raggaer** [comment]
When loading extensions, why `Errorf` and no `Fatalf`?

File: app/util/template.go [hunk 1]
```diff
@@ -87,6 +147,11 @@ func (t Tmpl) RenderWidget(req *http.Request, name string, args map[string]inter
 		if err := t.LoadTemplates("widgets/"); err != nil {
 			return nil, err
 		}
+
+		// Reload extension templates
+		if err := t.LoadExtensionTemplates("widgets"); err != nil {
+			Logger.Logger.Error(err.Error())
```
Comment (Raggaer) [inline app/util/template.go]: Maybe the errors here can be more verbose

**Comment by forgee** [comment]
I tried to view it from a user's perspective. Let's say you are new to Castro and install an extension made by someone else from the admin page. Then it turns out that the extension contains an error that prevents it from being loaded.

Now, if the app throws `Fatalf`, you can not start Castro at all to remove/disable the extension again. In order to be able to start Castro again, you would have to manually remove the entries from the database.

Using `Errorf` the extension will not work, but you can still start the app normally. This way you will be able to remove the extension again using the UI.

I think for someone who is not familiar with how the system works, it is a crucial difference.

Perhaps we could add another install status and have the UI mark the extension as "broken" when that happens, to make it more obvious. If you really think it should be `Fatalf` I will change it.

File: app/util/template.go (hunk 1 above)
Comment (forgee) [reply app/util/template.go]: I will fix this tomorrow.

**Comment by Raggaer** [comment]
Good point. I will try to review more changes this week

**Comment by Raggaer** [comment]
Any news on this @forgee ?

File: app/util/static.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// FileExists checks if the given resource exists
+func (e *StaticList) FileExists(url string) bool {
+	// Read lock mutex
+	e.rw.RLock()
+	defer e.rw.RUnlock()
+
+	// Loop list
+	for _, e := range e.list {
+
+		if strings.HasPrefix(strings.Replace(url, "/", "\\", -1), filepath.Join(e)) {
```
Comment (forgee) [inline app/util/static.go]: This only seems to work on Windows. Linux (and OSX?) uses forwardslashes in file paths. Possible fix:
`if strings.HasPrefix(url, filepath.Join(e)) || strings.HasPrefix(strings.Replace(url, "/", "\\", -1), filepath.Join(e)) {`

Comment (Raggaer) [reply app/util/static.go]: I pushed a new commit using `http.FileSystem`; Can you try if this is working? paths are always a headache

Comment (forgee) [reply app/util/static.go]: Working on Windows 7 and Ubuntu 14.04 :+1:

**Review by Raggaer — COMMENTED** [review]
This approach should work in all cases since theres no strings join magic

**Comment by Raggaer** [comment]
I accidentally merged branches

**Pull request merged**